File: windows-agent/LockedAgent/ScreenshotService.cs
using System.Drawing;
using System.Drawing.Imaging;
using System.Net.Http;
using System.Net.Http.Json;

namespace LockedAgent;

/// <summary>
/// Captures the screen the teacher's configured number of times per minute
/// (1-60), at random moments within each 60-second window rather than fixed
/// intervals, and uploads each frame to the backend as base64 JSON. Upload
/// failures are swallowed on purpose: a flaky network shouldn't crash the
/// agent mid-exam.
/// </summary>
public sealed class ScreenshotService : IDisposable
{
    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string _sessionId;
    private readonly ScreenshotConfig _config;
    private readonly Random _random = new();
    private readonly CancellationTokenSource _cts = new();
    private Task? _loopTask;

    public ScreenshotService(HttpClient http, string baseUrl, string sessionId, ScreenshotConfig config)
    {
        _http = http;
        _baseUrl = baseUrl.TrimEnd('/');
        _sessionId = sessionId;
        _config = config;
    }

    public void Start()
    {
        if (!_config.Enabled) return;
        _loopTask = RunAsync(_cts.Token);
    }

    private async Task RunAsync(CancellationToken token)
    {
        try
        {
            while (!token.IsCancellationRequested)
            {
                var windowStart = DateTime.UtcNow;

                var offsetsMs = new List<int>(_config.PerMinute);
                for (int i = 0; i < _config.PerMinute; i++)
                    offsetsMs.Add(_random.Next(0, 60_000));
                offsetsMs.Sort();

                foreach (var offsetMs in offsetsMs)
                {
                    var delay = windowStart.AddMilliseconds(offsetMs) - DateTime.UtcNow;
                    if (delay > TimeSpan.Zero) await Task.Delay(delay, token);
                    await CaptureAndUploadAsync();
                }

                var remaining = TimeSpan.FromSeconds(60) - (DateTime.UtcNow - windowStart);
                if (remaining > TimeSpan.Zero) await Task.Delay(remaining, token);
            }
        }
        catch (OperationCanceledException)
        {
            // Stopped by Dispose — expected when the exam ends or the student is excluded.
        }
    }

    private async Task CaptureAndUploadAsync()
    {
        try
        {
            using var bitmap = CaptureScreen();
            using var stream = new MemoryStream();
            bitmap.Save(stream, ImageFormat.Png);
            var imageBase64 = Convert.ToBase64String(stream.ToArray());

            await _http.PostAsJsonAsync($"{_baseUrl}/api/screenshot", new
            {
                sessionId = _sessionId,
                imageBase64,
            });
        }
        catch
        {
            // Best-effort upload; a dropped frame is not worth interrupting the exam for.
        }
    }

    private static Bitmap CaptureScreen()
    {
        int width = (int)System.Windows.SystemParameters.PrimaryScreenWidth;
        int height = (int)System.Windows.SystemParameters.PrimaryScreenHeight;
        var bitmap = new Bitmap(width, height);
        using var g = Graphics.FromImage(bitmap);
        g.CopyFromScreen(0, 0, 0, 0, new Size(width, height));
        return bitmap;
    }

    public void Dispose()
    {
        _cts.Cancel();
        _cts.Dispose();
    }
}

File: mac-agent/Sources/LockedAgentMac/ScreenshotService.swift
import Cocoa

/// Captures the screen the teacher's configured number of times per minute
/// (1-60), at random moments within each 60-second window rather than fixed
/// intervals — mirrors the Windows agent's ScreenshotService.cs exactly.
/// Requires the Screen Recording permission; silently captures nothing
/// (best-effort, matches the "don't interrupt the exam for a dropped frame"
/// posture already used for upload failures) if it isn't granted.
final class ScreenshotService {
    private let baseUrl: String
    private let sessionId: String
    private let authToken: String
    private let perMinute: Int
    private var task: Task<Void, Never>?

    init(baseUrl: String, sessionId: String, authToken: String, perMinute: Int) {
        self.baseUrl = baseUrl
        self.sessionId = sessionId
        self.authToken = authToken
        self.perMinute = perMinute
    }

    func start() {
        guard perMinute > 0 else { return }
        task = Task {
            while !Task.isCancelled {
                let windowStart = Date()
                let offsetsMs = (0..<perMinute).map { _ in Int.random(in: 0..<60_000) }.sorted()

                for offsetMs in offsetsMs {
                    let target = windowStart.addingTimeInterval(Double(offsetMs) / 1000)
                    let delay = target.timeIntervalSinceNow
                    if delay > 0 {
                        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                    }
                    if Task.isCancelled { break }
                    await captureAndUpload()
                }

                let remaining = 60.0 - Date().timeIntervalSince(windowStart)
                if remaining > 0 {
                    try? await Task.sleep(nanoseconds: UInt64(remaining * 1_000_000_000))
                }
            }
        }
    }

    func stop() {
        task?.cancel()
    }

    private func captureAndUpload() async {
        guard let image = CGWindowListCreateImage(.infinite, .optionOnScreenOnly, kCGNullWindowID, .bestResolution),
            let pngData = NSBitmapImageRep(cgImage: image).representation(using: .png, properties: [:])
        else {
            return
        }

        var request = URLRequest(url: URL(string: "\(baseUrl)/api/screenshot")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "sessionId": sessionId,
            "imageBase64": pngData.base64EncodedString(),
        ])
        _ = try? await URLSession.shared.data(for: request)
    }
}
